OSDOCS#12643: Docs for BYO OIDC auth provider (#93779)
openshift/openshift-docs: modules/external-auth-fields.adoc

|Configures the prefix that is applied to the cluster identity username attribute during the process of mapping JWT claims to cluster identity attributes. Must not be an empty string (`""`).

|`claimMappings.username.prefixPolicy`
|Configures how a prefix should be applied to the value of the JWT claim specified in the `claim` field. Allowed values are `Prefix`, `NoPrefix`, and omitted (not provided or an empty string).

When set to `Prefix`, the value specified in the prefix field is prepended to the value of the JWT claim. The prefix field must be set when `prefixPolicy` is `Prefix`.

When set to `NoPrefix`, no prefix is prepended to the value of the JWT claim.

When omitted, this means no opinion and the platform is left to choose any prefixes that are applied which is subject to change over time.

Currently, the platform prepends `{issuerURL}#` to the value of the JWT claim
when the claim is not `email`.

|`claimValidationRules`
|Configures the rules to be used by the Kubernetes API server for validating the claims in a JWT token issued by the identity provider. Validation rules are joined by an `AND` operation.

|`claimValidationRules.requiredClaim`

> |Configures the required claim and value that the Kubernetes API server will use to validate if an incoming JWT is valid for this identity provider.

Reviewer: Avoid future tense: 
```suggestion
|Configures the required claim and value that the Kubernetes API server uses to validate if an incoming JWT is valid for this identity provider.
```

> |Configures the required claim and value that the Kubernetes API server will use to validate if an incoming JWT is valid for this identity provider.

Reviewer: ```suggestion
|Configures the required claim and value that the Kubernetes API server uses to validate if an incoming JWT is valid for this identity provider.
```

|`claimValidationRules.requiredClaim.claim`
|Configures the name of the required claim. When taken from the JWT claims, the claim must be a string value. Must not be an empty string (`""`).

|`claimValidationRules.requiredClaim.requiredValue`

> |Configures the value that `claim` must have when taken from the incoming JWT claims. If the value in the JWT claims does not match, the token will be rejected for authentication. Must not be an empty string (`""`).

Reviewer: ```suggestion
|Configures the value that `claim` must have when taken from the incoming JWT claims. If the value in the JWT claims does not match, the token is rejected for authentication. Must not be an empty string (`""`).
```

|`claimValidationRules.type`
|Configures the type of the validation rule. Allowed values are `RequiredClaim` and omitted (not provided or an empty string).

> When set to `RequiredClaim`, the Kubernetes API server will be configured to validate that the incoming JWT contains the required claim and that its value matches the required value. The default value is `RequiredClaim`.

Reviewer: ```suggestion
When set to `RequiredClaim`, the Kubernetes API server is configured to validate that the incoming JWT contains the required claim and that its value matches the required value. The default value is `RequiredClaim`.
```

|`name`
|A required field that configures the unique human-readable identifier associated with the identity provider. It is used to distinguish between multiple identity providers and has no impact on token validation or authentication mechanics. Must not be an empty string (`""`).

|`oidcClients`
|Configures how on-cluster, platform clients should request tokens from the identity provider. Must not exceed 20 entries and entries must have unique namespace/name pairs.

|`oidcClients.clientID`
|Configures the client identifier, from the identity provider, that the platform component uses for authentication requests made to the identity provider. The identity provider must accept this identifier for platform components to be able to use the identity provider as an authentication mode. Must not be an empty string (`""`).

|`oidcClients.clientSecret`
|Configures the client secret used by the platform component when making authentication requests to the identity provider.

> When not specified, no client secret will be used when making authentication requests to the identity provider.

Reviewer: ```suggestion
When not specified, no client secret is used when making authentication requests to the identity provider.
```

> When specified, it references a secret in the `openshift-config` namespace that contains the client secret in the `clientSecret` key of the `.data` field. The client secret will be used when making authentication requests to the identity provider.

Reviewer: ```suggestion
When specified, it references a secret in the `openshift-config` namespace that contains the client secret in the `clientSecret` key of the `.data` field. The client secret is used when making authentication requests to the identity provider.
```

> Public clients do not require a client secret but private clients do require a client secret to work with the identity provider.

Reviewer: ```suggestion
Public clients do not require a client secret, but private clients do require a client secret to work with the identity provider.
```

// Module included in the following assemblies:
//
// * authentication/external-auth.adoc

:_mod-docs-content-type: PROCEDURE
[id="external-auth-fields_{context}"]
= OIDC provider configuration parameters

> // Note: Because this is tech preview, the API docs won't contain this information, so we're including this table now. Consider whether to continue including this table or not once it's GA and in the API docs.

Reviewer 2: @bergerhoffer , fine for me.

|`claimMappings.extra`
|An optional field for configuring the mappings used to construct the extra attribute for the cluster identity. When omitted, no extra attributes will be present on the cluster identity. Key values for extra mappings must be unique. A maximum of 64 extra attribute mappings may be provided.

|`claimMappings.extra.key`
a|A required field that specifies the string to use as the extra attribute key, in the format of <domain>/<path>, for example `example.org/test`. The key must not exceed 510 characters and must not be empty.

The domain portion of the key must be a valid RFC1123 subdomain and have the following characteristics:

* It must not exceed 253 characters in length.
* It must start and end with an alphanumeric character.
* It must only contain lower case alphanumeric characters and `-` or `.`.
* It must not use the reserved domains, or be subdomains of, `kubernetes.io`, `k8s.io`, and `openshift.io`.

The path portion of the key must have the following characteristics:

* Must not be empty.
* Must consist of at least one alphanumeric character, percent-encoded octets, `-`, `.`, `_`, `~`, `!`, `$`, `&`, `'`, `(`, `)`, `*`, `+`, `,`, `;`, `=`, and `:`.

> // TODO: Clarify if this means it needs at least one alphanumeric character and one of these symbols?

Reviewer 3: @xingxingxia if this documentation is targeted for TP, then including this is fine as the fields will show up in both OCP and HCP in TP. That being said, there is no business logic behind the fields for OCP yet, so setting them will do nothing for now.

@bergerhoffer the at least one of is just that the value provided by a user must contain at least one of those above characters. The `and` in the statement may be a bit misleading, and that is something we can change to `or` if it would read better.

> // TODO: Clarify if this means it needs at least one alphanumeric character and one of these symbols?

Author: Edit to add that I noticed now that the PR only went in for 4.20, so the descriptions will remain as is in the API for 4.19. But I'll still use the new ones from the PR for my doc.

Precisely one of `claim` or `expression` must be set. When specified, `expression` must be at least 1 character in length and must not exceed 4096 characters in length.

|`claimMappings.username`
|A name of the claim that should be used to construct usernames for the cluster identity. The default is `sub`.

|`claimMappings.username.claim`
|A JWT token claim to be used in the mapping.

|`claimMappings.username.prefix`
|How to prefix the value from the token in the result of the claim mapping.
// TODO: This had no description, so I made one based on group

|`claimMappings.username.prefix.prefixString`
|A string to prefix the value from the token in the result of the claim mapping.
// TODO: This had no description, so I made one based on group

|`claimMappings.username.prefixPolicy`
|How a prefix should apply. Allowed values are `""`, `NoPrefix`, or `Prefix`. By default, claims other than `email` will be prefixed with the issuer URL to prevent naming clashes with other plugins. Set to `NoPrefix` to disable prefixing.

> // TODO: Does this mean that if you set it to "" that it prefixes with the issuer URL? and if you want to set your own, you set it to Prefix? And set to NoPrefix to do no prefix at all?

Author: Thanks, updated!

|`claimValidationRules`
|Rules that are applied to validate token claims to authenticate users.

|`claimValidationRules.requiredClaim`
|Allows configuring a required claim name and its expected value.

|`claimValidationRules.requiredClaim.claim`
|A name of a required claim. Only claims with string values are supported.

|`claimValidationRules.requiredClaim.requiredValue`
|The required value for the claim.

|`claimValidationRules.type`
|The type of validation rule.

> // TODO not sure what to do with: enum: RequiredClaim

Author: @liouk @xingxingxia I see that it lists an enum of RequiredClaim, but I'm unsure what that means, if there's only one option listed.

Reviewer 2: @bergerhoffer , yes only one option. I had a case (OCP-79806 Updating oidcProviders should work well in OCP) in our test case system where it covered `Test "claimValidationRules"` and `Update "requiredValue"`, which should clearly demo what it means.

Reviewer 3: https://github.com/openshift/api/blob/a463917701ab4506585893739e576ffd1e6b924e/config/v1/types_authentication.go#L711-L720 might also help

|`issuer`
|A required field that describes attributes of the OIDC token issuer.

> // TODO: typo "atributes" in code

Author: @liouk Just FYI that there was a typo of "atributes" in the code that you might want to fix. I'm also happy to take a look through all of the descriptions for 4.20 to make some suggestions.

Reviewer 2: https://issues.redhat.com/browse/OCPBUGS-56851

Author: @xingxingxia I commented on your Jira, but I believe @everettraven's PR https://github.com/openshift/api/pull/2314 already fixed this and the other things you noted.

|`issuer.audiences`
|A required field that specifies an array of audiences that the token was issued for. Valid tokens must include at least one of these values in their `aud` claim. Must be set to exactly one value.

> // TODO: Why does it say that it must be set to exactly one value? examples in test cases have multiple

Reviewer 3: This has been updated recently - https://github.com/openshift/api/blob/a463917701ab4506585893739e576ffd1e6b924e/config/v1/types_authentication.go#L260-L264
